### PR TITLE
Data, expressions, statements and simple generation.

### PR DESCRIPTION
**Pull request by Vsevolod-Livinskij**

### CODE REVIEW

File: src/expr_test.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+int main() {
+    IRValue start_val (IntTypeID::INT);
+    start_val.getValueRef<int32_t>() = 0;
+    start_val.setIsUndefined(false);
+    auto start_expr = std::make_shared<ConstantExpr>(start_val);
+
+   IRValue end_val (IntTypeID::INT);
+   end_val.getValueRef<int32_t>() = 0;
+   end_val.setIsUndefined(false);
+   auto end_expr = std::make_shared<ConstantExpr>(end_val);
```
Comment (dbabokin) [inline src/expr_test.cpp]: Wrong indentation. Make sure you apply clang-format.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -16,7 +16,7 @@
 #
 ###############################################################################
 
-set(LIB_SRCS type.cpp ir_value.cpp enums.h hash.cpp hash.h)
+set(LIB_SRCS type.cpp ir_value.cpp enums.h hash.cpp hash.h data.cpp data.h)
```
Comment (dbabokin) [inline src/CMakeLists.txt]: Let's have files in alphabetical order, one per line.
Like here: https://github.com/ispc/ispc/blob/master/CMakeLists.txt#L218

File: src/utils.h [hunk 3]
```diff
@@ -28,3 +32,30 @@ limitations under the License.
...
+        std::uniform_int_distribution<long long> dis(from, to);
+        return dis(rand_gen);
+    }
+
+  private:
+    uint64_t seed;
+    std::mt19937_64 rand_gen;
+};
+
+template <> inline bool RandValGen::getRandValue<bool>(bool from, bool to) {
```
Comment (dbabokin) [inline src/utils.h]: Does it make sense to accept any parameters for bool version?

File: src/utils.h [hunk 4]
```diff
@@ -28,3 +32,30 @@ limitations under the License.
...
     } while (false)
+
+// According to the agreement, Random Value Generator is the only way to get any random value in YARPGen. It is used for different random decisions all over the source code.
+class RandValGen {
+  public:
+    // Specific seed can be passed to constructor to reproduce the test.
+    // Zero value is reserved (it notifies RandValGen that it can choose any)
+    RandValGen(uint64_t _seed);
+
+    template <typename T> T getRandValue(T from, T to) {
```
Comment (dbabokin) [inline src/utils.h]: Should we have a version, which doesn't take any parameters, but returns the value in the whole type range?

File: src/context.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "data.h"
+#include <map>
+#include <string>
+
+namespace yarpgen {
+
+// Class that is used to determine the evaluation context.
+// It allows us to evaluate the same arithmetic tree with different input values.
+class EvalCtx {
```
Comment (dbabokin) [inline src/context.h]: Not sure that I understand the idea here, but ok as it's just a placeholder for now.

File: src/utils.h (hunk 3 above)
Comment (Vsevolod-Livinskij) [reply src/utils.h]: It is explicit template instantiation for function. Let's keep them uniform.

File: src/utils.h (hunk 4 above)
Comment (Vsevolod-Livinskij) [reply src/utils.h]: Maybe. We also need to think about a version that accepts IRValue.

File: src/context.h (hunk 5 above)
Comment (Vsevolod-Livinskij) [reply src/context.h]: The idea here is to be able to evaluate the same expression with different values without actually changing them. It allows us to override them within one arithmetic expression.

File: src/utils.h (hunk 4 above)
Comment (dbabokin) [reply src/utils.h]: Make sense, you can add TODO for that.

File: src/data.h [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    //TODO:
+    // We want elements of the array to have different values.
+    // It is the only way to properly test masked instructions and optimizations
+    // designed to work with them.
+    // Each vector represents a "cluster" with different values that is "copied"
+    // over the whole array. The size of the cluster should be smaller than the
+    // typical target architecture vector size. This way we can cover
+    // all of the interesting cases while preserving the simplicity of
+    // the analysis.
+    std::shared_ptr<Data> init_vals;
```
Comment (dbabokin) [inline src/data.h]: I'm not sure that it makes sense to recurse to Data, but not to go directly to IRValue (or set of IRValues).

File: src/CMakeLists.txt (hunk 2 above)
Comment (Vsevolod-Livinskij) [reply src/CMakeLists.txt]: Done

File: src/data.h (hunk 6 above)
Comment (Vsevolod-Livinskij) [reply src/data.h]: What about an array of structures? The IRValue represents a constant of integral type.

File: src/utils.h (hunk 4 above)
Comment (Vsevolod-Livinskij) [reply src/utils.h]: Done

File: src/stmt.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
+/*
+Copyright (c) 2015-2019, Intel Corporation
```
Comment (dbabokin) [inline src/stmt.cpp]: If it's brand new code, it should be 2020.

File: src/data.h (hunk 6 above)
Comment (dbabokin) [reply src/data.h]: Makes sense, but it should probably be spelled out if this is a single element or something else.

File: src/stmt.cpp [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        new_offset.erase(new_offset.size() - 4, 4);
+    }
+}
+
+std::shared_ptr<LoopNestStmt> LoopNestStmt::generateStructure(std::shared_ptr<GenCtx> ctx) {
+    auto gen_pol = ctx->getGenPolicy();
+    size_t nest_depth = rand_val_gen->getRandId(gen_pol->loop_nest_depth_distr);
+
+    // We need to limit a maximal loop depth that we can generate
+    nest_depth = std::min(gen_pol->loop_depth_limit - ctx->getLoopDepth(), nest_depth);
```
Comment (dbabokin) [inline src/stmt.cpp]: VS2017 build:
C:\projects\yarpgen\src\stmt.cpp(230): error C2039: 'min': is not a member of 'std'

**Pull request merged**